smartsim/_core/_install/utils.py

```diff
@@ -0,0 +1,161 @@
+        **kwargs: t.Any,
+    ) -> None:
+        """Download and decompress a package
+
+        :param source: URL to a particular package
+        :param destination: Where to unpack the archive
+        """
+        local_file, _ = urlretrieve(source, **kwargs)
+        cls._from_local_archive(local_file, destination)
+        os.remove(local_file)
+
+    @staticmethod
+    def _from_git(
+        source: str, destination: pathlib.Path, **clone_kwargs: t.Any
+    ) -> None:
+        """Clone a repository
+
+        :param source: Path to the remote (URL or local) repository
+        :param destination: where to clone the repository
+        :param clone_kwargs: various options to send to the clone command
+        """
+        is_mac = OperatingSystem.autodetect() == OperatingSystem.DARWIN
+        is_arm64 = Architecture.autodetect() == Architecture.ARM64
+        if is_mac and is_arm64:
+            config_options = ["--config core.autocrlf=false", "--config core.eol=lf"]
+        else:
+            config_options = None
+        git.Repo.clone_from(
+            source, destination, multi_options=config_options, **clone_kwargs
+        )
+
+    @classmethod
+    def retrieve(
+        cls, source: _PathLike, destination: pathlib.Path, **retrieve_kwargs: t.Any
+    ) -> None:
+        """Primary method for retrieval
+
+        Automatically choose the correct method based on the extension and/or source
+        of the archive. If downloaded, this will also decompress the archive and
+        extract
+
+        :param source: URL or path to find the package
+        :param destination: where to place the package
+        :raises UnsupportedArchive: Unknown archive type
+        :raises PathNotFound: Path to archive does not exist
+        """
+        url_scheme = urlparse(str(source)).scheme
+        if str(source).endswith(".git"):
+            cls._from_git(str(source), destination, **retrieve_kwargs)
+        elif url_scheme == "http":
+            cls._from_http(str(source), destination, **retrieve_kwargs)
+        elif url_scheme == "https":
+            cls._from_http(str(source), destination, **retrieve_kwargs)
+        else:  # This is probably a path
+            source_path = pathlib.Path(source)
+            if source_path.exists():
+                if source_path.is_dir():
+                    cls._from_local_directory(source, destination, **retrieve_kwargs)
+                elif source_path.is_file() and source_path.suffix in (
+                    ".gz",
+                    ".zip",
+                    ".tgz",
+                ):
+                    cls._from_local_archive(source, destination, **retrieve_kwargs)
+                else:
+                    raise UnsupportedArchive(
+                        f"Source ({source}) is not a supported archive or directory "
+                    )
+            else:
+                raise PathNotFound(f"Package path or file does not exist: {source}")
```

Comment:
super minor style thing: Consider inverting your error checks to avoid deeply nesting:
```suggestion
            if not source_path.exists():
                raise PathNotFound(f"Package path or file does not exist: {source}")
            if source_path.is_dir():
                cls._from_local_directory(source, destination, **retrieve_kwargs)
            elif source_path.is_file() and source_path.suffix in (
                ".gz",
                ".zip",
                ".tgz",
            ):
                cls._from_local_archive(source, destination, **retrieve_kwargs)
            else:
                raise UnsupportedArchive(
                    f"Source ({source}) is not a supported archive or directory "
                )
```

Reply:
if you want, I would also consider making the `elif` blocks just `if` blocks with early returns to avoid the need to nest on the `else` clause here